### ISSUE TITLE
FInal token index when training a model

### ISSUE THREAD

**Issue by robertodessi** [rmokady/CLIP_prefix_caption#42]
Hi, thanks a lot for making the code available, it's a great resource to use!

I was wondering why the index when computing the loss from the output of gpt is shifted by 1 on the left:
https://github.com/rmokady/CLIP_prefix_caption/blob/main/train.py#L315

shouldn't it be `logits = outputs.logits[:, dataset.prefix_length :]`

Thanks!

**Comment by Twilighter9527** [comment]
when I follow you change logits to be "logits = outputs.logits[:, dataset.prefix_length :]",the loss become very slow quickly and the inference caption become "\n\n\n\n.......", do you have this situation？or that should be "logits = outputs.logits[:, dataset.prefix_length - 1: -1]"thank you.
![image](https://user-images.githubusercontent.com/49281157/174536599-e4d59f99-c897-41ac-8d44-de8e769378c3.png)

**Comment by baaaad** [comment]
the gpt2 acts the ''**next token predixtion**', so the token in time step t is predicted by the token in time step t-1. the final token does not need to predict for the next token. so it is right to use **logits = outputs.logits[:, dataset.prefix_length -1 : -1]**

**Comment by Twilighter9527** [comment]
you can see the [source](https://github.com/huggingface/transformers/blob/v4.20.1/src/transformers/models/gpt2/modeling_gpt2.py#L947)
![image](https://user-images.githubusercontent.com/49281157/175273368-a31ca4b1-11e4-45f5-bd02-90b265ead463.png)

**Comment by robertodessi** [comment]
True, makes sense thanks!
